### ISSUE TITLE
Missing -fPIC flag

### ISSUE THREAD

**Issue by yurivict** [Z3Prover/z3#2130]
The build of the python module fails on FreeBSD 12: http://beefy11.nyi.freebsd.org/data/head-i386-default/p492480_s343922/logs/py27-z3-solver-4.7.1.log

```
ld: error: can't create dynamic relocation R_386_32 against symbol: .L.str.46 in readonly segment; recompile object files with -fPIC or pass '-Wl,-z,notext' to allow text relocations in the output
```

Adding this flag to this line https://github.com/Z3Prover/z3/blob/master/src/api/python/setup.py#L19 doesn't pass it to the build. Adding it to the ```CXXFLAGS``` environment variable also doesn't help.

IMO, it is a bug that it ignores the ```CXXFLAGS``` setting in setup.py#L19 .

Please advise how to fix this problem.

**Comment by NikolajBjorner** [comment]
@rhelmot: The script requires some command-line arguments. Is there a cheat sheet that would let me run it? See https://github.com/Z3Prover/z3/blob/master/scripts/mk_nuget_release.py where we pull a latest release and create a package. Do you know if we can do something similar for Pip packaging to get it streamlined and up to date with current releases?

@yurivict: There is a comment in the script that says "Compiliation, or installation, please submit issues to https://github.com/angr/angr-z3", but it is useful to see this issue as it directed my attention to this script. The 4.7.1 release is a bit old.

**Comment by rhelmot** [comment]
- the most recent release on pypi is [4.8.0.0](https://pypi.org/project/z3-solver/), though someone just asked me to update it to .2 a few days ago. 4.7.1 is probably because angr has a pinned dependency on that version bc 4.8 changed some APIs and I haven't really wanted to fix it.
- by "the script" are you referring to setup.py? That's a super standard python packaging file, and it has its own usage information: `python setup.py --help` will tell you most things you care about, `python setup.py --help-commands` will tell you the rest of them.
- If you're referring to the super-build-script in angr/angr-dev linked at the bottom of the script, that doesn't take command line arguments (yet?), you just comment and uncomment the build rules near the top
- I will submit a fix for this today and update the pypi package to 4.8.5.0.

**Comment by NikolajBjorner** [comment]
The "script" is setup.py. I can see it has bdist_egg listed, but bdist_wheel somehow also calls into build scripts. There is some indication that this is taking over what is called "egg". Without having background in wheels and eggs it is somewhat opaque to me.

 Instead of building z3 from scratch in this script could it download latest release and package a wheel for various platforms?

**Comment by rhelmot** [comment]
Truthfully I do not understand the machinations of setup.py. it is deep magic that only the setuptools devs truly understand, and yet it is required for packaging any python script.

Isn't the point of a build script to allow customization of builds? I usually think that if a user doesn't want do do a build they should download a binary distribution from a package manager.

**Comment by yurivict** [comment]
> Truthfully I do not understand the machinations of  setup.py. it is deep magic that only the setuptools devs truly understand, and yet it is required for packaging any python script.

This feature is great for their job security.

**Comment by rhelmot** [comment]
er, I should have actually answered your question: "eggs" are a pretty deprecated mechanism of python packaging so far as distributions go, but they are still used internally during the build process and for record keeping in installations.

Generally, my lexicon for using setup.py is:

- `pip install .` for doing a normal installation
- `pip install -e .` for doing a build and symlinking the result into my current python environment
- `python setup.py build` for updating a build done in the previous manner
- `python setup.py bdist_wheel` for building a binary distribution (into the `dist` directory)
- `python setup.py sdist` for building a source distribution

**Comment by NikolajBjorner** [comment]
It seems I am asking the wrong type of questions. The background is that I would like to streamline the creation of pip installations such that z3-solver can be updated when there is a new version of Z3. I am increasingly getting hints that users just assume there is some pip package for z3, which there is, if they can figure out that the "z3" pip package is for zfs on s3 and not about our Z3 and then they can find one of the two pip distributions, where z3-solver is the most up-to-date version.

**Comment by rhelmot** [comment]
@yurivict I suspect the reason that adding -fPIC to that line didn't affect anything was because it actually needs to be in the environment at configure-time, and the python build script will reuse the previous configuration run. To address this I added the capability for `python setup.py clean` to totally wipe out the build directory and start over, see #2133

**Comment by yurivict** [comment]
@rhelmot I just updated the port to the latest version. Will see if it will be failing the same way, and will add this flag into the environment variables passed to the configure stage. It began failing in the central build a week ago for no apparent reason. It was building fine before for many months.

**Comment by yurivict** [comment]
Regarding packaging, ```pip``` packaging is python-specific. All modern systems have their own packaging systems that are language-agnostic. Language-agnostic packaging is obviously preferable. You need to request that packages are created for ```z3-solver``` in all systems, so that ```z3-solver``` would be like ```numpy```'s python package that is available everywhere: https://repology.org/metapackage/python:numpy/versions

Currently, z3-solver is only available on FreeBSD and Arch Linux. There also need to also be Fedora, Debian, MacOS, Gentoo packages. Each one has a mailing list, with people likely willing to help with this.

**Comment by rhelmot** [comment]
Here's how I do releases:

- release a source distribution. that is, `python setup.py sdist` (which will bundle up all the source and python binding files) and then `twine upload dist/z3-solver*.tar.gz`
- build the linux binary distributions. In order to get around the fact that "linux" is not a monolith and we need to build with maximum compatibility, I use [a special docker image](https://github.com/pypa/manylinux) that lets us build with compatibility for centos 5 (!!), which should deeeefinitely be compatible with all modern linux distros.
- I have a [script](https://github.com/angr/angr-dev/blob/master/admin/bdist.py) which automates bits of this process. To build z3 copied the current pypi source release, first wait a few minutes for the pypi caches to clear, then uncomment only the `Target` line which references `pypi_target='z3-solver'. The script will ask for sudo in order to run docker. It will put its results in a folder named `bdist`
- rename the `py2` or `py3` in the filenames to `py2.py3` in order to indicate that the file you built works on both python versions. I have never figured out how to indicate in the build configuration that a build will work for both python versions but _not_ any old platform. Fortunately this metadata is only stored in the filename (...)
- `twine upload` the wheels, there should be two of them, for i386 and x86_64
- the script should work on windows as well. You will need to set the `PYTHON_32` and `PYTHON_64` environment variables to point to the 32-bit and 64-bit python executables, respectively
- same deal with renaming
- same deal with twine upload

at the end of the day, there should be 5 files at https://pypi.org/project/z3-solver/#files: the source, the two linux wheels, and the two windows wheels. I don't have a good way to build for macos.

If at any point you mess up, you need to increment the version number by adding `env={'Z3_VERSION_SUFFIX': '.post1'}` or post2 etc to the Target line and starting over. pypi has very very strict rules forbidding deleting and reuploaing files.

**Comment by NikolajBjorner** [comment]
@levnach

**Comment by LordAro** [comment]
How goes the update process on this? I notice pypi is still on 4.8.0...

**Comment by NikolajBjorner** [comment]
It should be possible to put this into an Azure Pipeline and have it as push button release.
Current pipelines are now visible at:
https://z3build.visualstudio.com/Z3Build/_build?definitionId=10
The infrastructure supports the three main platforms.
What I would need is converting the steps indicated above as a sequence of executable commands that can be put into a yaml file. If the centos build is a strong requirement, then the process would require docker integration, which is possible, but less simple. It also seems reasonable if the pip package could use sources and binaries based on the releases directly instead of building them on its own. The github repository contains tagged releases and I use them to create nuget packages. Is this possible for pip?

**Comment by rhelmot** [comment]
You can repackage any source release into a python sdist by just downloading it and running `python setup.py sdist` from the bundled setup.py.

I'll submit a PR in a few minutes for repackaging a binary release into a python wheel :)

The centos build should really happen; it is the standard for python binary distributions. The repo I linked for the docker image is run by the pypa themselves! The problem is that there's no way for a python distribution to specify which linux distro it targets (automatically detecting distro is [disgustingly hard](https://bugs.python.org/issue1322)) it turns out). The [standard](https://www.python.org/dev/peps/pep-0513/) is to conform to a specific set of dynamic linking requirements, which are largely satisfied when you build in centos 5. Technically, z3's link against libgomp makes it not a standards-compliant manylinux1 wheel, which is a little sketchy. The advice is that we bundle it in the distribution but tbh until I actually see someone with a problem related to libgomp I don't want to deal with it. I can't seem to find the issue where this was discussed earlier but we should probably have it linked here for reference.

**Comment by NikolajBjorner** [comment]
> Technically, z3's link against libgomp makes it not a standards-compliant manylinux1 wheel, which is a little sketchy.
We don't need to link against libgomp any longer. The dependency is deprecated.

**Comment by NikolajBjorner** [comment]
can we close this with 4.8.6 on pypi?

**Comment by rhelmot** [comment]
Yep, really it could have been closed with b702cad being merged :)
